### DISCUSSION TITLE
how to set the writeableTopics

### DISCUSSION THREAD

**Discussion by tasisale** [tillsteinbach/WeConnect-mqtt#64] [Q&A]
i don't understand how i can set any of the new -> writeableTopics
can you give me a example?

regards
simon

**Comment by tillsteinbach** [comment]
To use a writeable topic you publish to the respective topic.

For e.g. temperatures you just send the value.
To control climatization or charging you send „start“ or „stop“ to the control topic.

**Reply by tasisale** [reply]
Sorry, can you give me a example via fhem or so!??!

**Comment by tillsteinbach** [comment]
Sorry, for FHEM you have to check the manual it is probably something like

```
attr Devicename setList climatisation_on weconnect/0/vehicles/YOURVIN/controls/climatisation start\
climatisation_off weconnect/0/vehicles/YOURVIN/controls/climatisation stop
```

**Reply by tasisale** [reply]
ok, thx, will work, bite i get an error, like this.
![error_json_start-stop](https://user-images.githubusercontent.com/78930595/149796799-aad8a31c-0199-4945-8aa6-ef3c65bf4475.jpg)
--
![error_json_temperatur](https://user-images.githubusercontent.com/78930595/149797375-b104ac13-4d9b-48b7-8afd-0c6c3dbfec51.jpg)

**Reply by tillsteinbach** [reply]
This looks like a FHEM error. But there is a good chance that your command worked as the callback was reported back

**Reply by boredomwontgetus** [reply]
To me this looks like you configured FHEM to evaluate `.../climatisation/climatisationSettings/request/1/body` or such a topic.
What does your read and set List attributes look like?

**Reply by tasisale** [reply]
yes it works, but i start the climatisation and the targetTemperature_C will change to 20.5, he will not use the installed 25.

**Reply by tasisale** [reply]
> To me this looks like you configured FHEM to evaluate `.../climatisation/climatisationSettings/request/1/body` or such a topic. What does your read and set List attributes look like?

`<html><body>


setList | climatisation_on weconnect/vehicles/WVWZZZAUZLW916270/controls/climatisation start
-- | --



</body>
</html>`
`readingList -> myBroker:weconnect/vehicles/WVWZZZAUZLW916270/controls/climatisation:.* controlsclimatisation`

**Reply by boredomwontgetus** [reply]
and readlist?
The message you are seeing....what topic is it published on? can you show the output of the raw mqtt-stream after you set clima to start/stop without fhem in the middle?

**Reply by tasisale** [reply]
`readingList -> myBroker:weconnect/vehicles/WVWZZZAUZLW916270/controls/climatisation:.* controlsclimatisation`

**Reply by tasisale** [reply]
this is display, i start via fhem
![error_json_start-only](https://user-images.githubusercontent.com/78930595/149799681-c9ab4798-fc4d-437b-8b50-094ae637fd3e.jpg)
![Foto 22-01-17 16-40-39 5558](https://user-images.githubusercontent.com/78930595/149799833-6939ead8-662d-4ecb-a616-ad40721a6be0.png)

**Reply by boredomwontgetus** [reply]
i am sorry...i have to admit....it don't understand fully.
this does not look like your full readinglist. it just looks like an excerpt. 

lets try this: can your provide your raw definition of your device.

and: The message you are seeing....what topic is it published on? can you show the output of the raw mqtt-stream after you set clima to start/stop without fhem in the middle?

**Reply by tillsteinbach** [reply]
> yes it works, but i start the climatisation and the targetTemperature_C will change to 20.5, he will not use the installed 25.

Oh, yes, you found a bug there.

**Reply by tasisale** [reply]
> > yes it works, but i start the climatisation and the targetTemperature_C will change to 20.5, he will not use the installed 25.
> 
> Oh, yes, you found a bug there.

thx for you answerd.
i will test the next version after you will fix it.

**Comment by tillsteinbach** [comment]
I tried to start a wiki page for FHEM: https://github.com/tillsteinbach/WeConnect-mqtt/wiki/FHEM-examples

**Comment by tasisale** [comment]
after the last update i will get an error to set the climatisation start:
`Error setting value: id /vehicles/WVWZZZAUZLW916270/controls/climatisation cannot be set to value ControlOperation.NONE. You need to provide it in the correct format select one of [start, stop]`

**Reply by tasisale** [reply]
`climatisation_on weconnect/0/vehicles/WVWZZZAUZLW916270/controls/climatisation start`

**Reply by tillsteinbach** [reply]
Can you try:
`attr MyCar setList climatisation:start,stop weconnect/0/vehicles/AAABBBCCCDD123456/controls/climatisation $EVTPART1\
charging:start,stop weconnect/0/vehicles/AAABBBCCCDD123456/controls/charging $EVTPART1`

**Reply by tillsteinbach** [reply]
Error should now be gone in 0.27.0

**Reply by tasisale** [reply]
ok, i will test tomorrow.

thx and a good night

**Reply by tasisale** [reply]
now i will get a lot of messages:
`request_0
request_0_operation
request_0_requestId
request_0_status
...
request_14
request_14_operation
request_14_requestId
request_14_status
...`
all with no data, return.
at the moment from 0 to 62?
what is this?

**Comment by tasisale** [comment]
with the new version, it will not work, sorry.
journalctl -u weconnect-mqtt says:
`Jan 23 22:27:00 srvfhem python3[210503]: 2022-01-23T22:27:00+0100:ERROR:client:Caught exception in on_message: payload must be a string, bytearray, int, float or None.
Jan 23 22:27:00 srvfhem python3[210503]: Exception in thread Thread-2:
Jan 23 22:27:00 srvfhem python3[210503]: Traceback (most recent call last):
Jan 23 22:27:00 srvfhem python3[210503]:   File "/home/fhemadmin/.local/lib/python3.8/site-packages/weconnect/addressable.py", line 344, in value
Jan 23 22:27:00 srvfhem python3[210503]:     raise ValueError('Value is not in allowed values')
Jan 23 22:27:00 srvfhem python3[210503]: ValueError: Value is not in allowed values
Jan 23 22:27:00 srvfhem python3[210503]: The above exception was the direct cause of the following exception:
Jan 23 22:27:00 srvfhem python3[210503]: Traceback (most recent call last):
Jan 23 22:27:00 srvfhem python3[210503]:   File "/home/fhemadmin/.local/lib/python3.8/site-packages/weconnect_mqtt/weconnect_mqtt_base.py", line 552, in on_message_callback
Jan 23 22:27:00 srvfhem python3[210503]:     attribute.value = msg.payload.decode()
Jan 23 22:27:00 srvfhem python3[210503]:   File "/home/fhemadmin/.local/lib/python3.8/site-packages/weconnect/addressable.py", line 361, in value
Jan 23 22:27:00 srvfhem python3[210503]:     raise ValueError(f'id {self.getGlobalAddress()} cannot be set to value {newValue}.'
Jan 23 22:27:00 srvfhem python3[210503]: ValueError: id /vehicles/WVWZZZAUZLW916270/controls/charging cannot be set to value ControlOperation.NONE. You need to provide it in the correct format select one of [start, stop]
Jan 23 22:27:00 srvfhem python3[210503]: During handling of the above exception, another exception occurred:
Jan 23 22:27:00 srvfhem python3[210503]: Traceback (most recent call last):
Jan 23 22:27:00 srvfhem python3[210503]:   File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
Jan 23 22:27:00 srvfhem python3[210503]:     self.run()
Jan 23 22:27:00 srvfhem python3[210503]:   File "/usr/lib/python3.8/threading.py", line 870, in run
Jan 23 22:27:00 srvfhem python3[210503]:     self._target(*self._args, **self._kwargs)
Jan 23 22:27:00 srvfhem python3[210503]:   File "/home/fhemadmin/.local/lib/python3.8/site-packages/paho/mqtt/client.py", line 3591, in _thread_main
Jan 23 22:27:00 srvfhem python3[210503]:     self.loop_forever(retry_first_connection=True)
Jan 23 22:27:00 srvfhem python3[210503]:   File "/home/fhemadmin/.local/lib/python3.8/site-packages/paho/mqtt/client.py", line 1756, in loop_forever
Jan 23 22:27:00 srvfhem python3[210503]:     rc = self._loop(timeout)
Jan 23 22:27:00 srvfhem python3[210503]:   File "/home/fhemadmin/.local/lib/python3.8/site-packages/paho/mqtt/client.py", line 1164, in _loop
Jan 23 22:27:00 srvfhem python3[210503]:     rc = self.loop_read()
Jan 23 22:27:00 srvfhem python3[210503]:   File "/home/fhemadmin/.local/lib/python3.8/site-packages/paho/mqtt/client.py", line 1556, in loop_read
Jan 23 22:27:00 srvfhem python3[210503]:     rc = self._packet_read()
Jan 23 22:27:00 srvfhem python3[210503]:   File "/home/fhemadmin/.local/lib/python3.8/site-packages/paho/mqtt/client.py", line 2439, in _packet_read
Jan 23 22:27:00 srvfhem python3[210503]:     rc = self._packet_handle()
Jan 23 22:27:00 srvfhem python3[210503]:   File "/home/fhemadmin/.local/lib/python3.8/site-packages/paho/mqtt/client.py", line 3033, in _packet_handle
Jan 23 22:27:00 srvfhem python3[210503]:     return self._handle_publish()
Jan 23 22:27:00 srvfhem python3[210503]:   File "/home/fhemadmin/.local/lib/python3.8/site-packages/paho/mqtt/client.py", line 3327, in _handle_publish
Jan 23 22:27:00 srvfhem python3[210503]:     self._handle_on_message(message)
Jan 23 22:27:00 srvfhem python3[210503]:   File "/home/fhemadmin/.local/lib/python3.8/site-packages/paho/mqtt/client.py", line 3570, in _handle_on_message
Jan 23 22:27:00 srvfhem python3[210503]:     on_message(self, self._userdata, message)
Jan 23 22:27:00 srvfhem python3[210503]:   File "/home/fhemadmin/.local/lib/python3.8/site-packages/weconnect_mqtt/weconnect_mqtt_base.py", line 559, in on_message_callback
Jan 23 22:27:00 srvfhem python3[210503]:     self.publish(topic=msg.topic, qos=1, retain=False, payload=oldValue)
Jan 23 22:27:00 srvfhem python3[210503]:   File "/home/fhemadmin/.local/lib/python3.8/site-packages/paho/mqtt/client.py", line 1247, in publish
Jan 23 22:27:00 srvfhem python3[210503]:     raise TypeError(
Jan 23 22:27:00 srvfhem python3[210503]: TypeError: payload must be a string, bytearray, int, float or None.
`

**Reply by tillsteinbach** [reply]
Grrr, yes a 0.27.1 was necessary

**Reply by tasisale** [reply]
yes, perfect work.
it works now again.
But what is this -> `request_0 request_0_operation request_0_requestId request_0_status ... request_14 request_14_operation request_14_requestId request_14_status`

it shows as the changes from the WeConnect - APP will make a new request entry via mqtt without value.

**Reply by tillsteinbach** [reply]
This should not happen if you only send one command at a time.

**Reply by tasisale** [reply]
i have test again, will set temp via the app, and the result via mqtt shows ;-(!
![mqtt_add-when-set-via app](https://user-images.githubusercontent.com/78930595/150757662-60a89253-9b0d-4d5b-b0f3-6dc5e2dcf1e3.jpg)

**Reply by tillsteinbach** [reply]
Looks like the App generates 7 requests when you change settings. You can just ignore those

**Reply by tasisale** [reply]
ok, but it will be set at every changes via app an new -> topic via MQTT, show this:
![reqests](https://user-images.githubusercontent.com/78930595/150759836-a4cfee1b-3383-4242-ae33-777220895439.jpg)

**Reply by tillsteinbach** [reply]
hope some of the problems are fixed now in 0.28.0. MQTT is driving me crazy! The protocol is so limited!

Also there is an update of the FHEM wiki page: https://github.com/tillsteinbach/WeConnect-mqtt/wiki/FHEM-examples

**Comment by boredomwontgetus** [comment]
With the latest version starting/stoping charging does not seam to work for me.

If i do 
```mosquitto_pub -h my.broker.net -u user -P pwd -t weconnect/0/vehicles/WVWXXXXXX/controls/charging_writetopic -m start```
nothing ever happens. 

but using weconnect-cli like so:
```weconnect-cli --username weconnectuser --password weconnectpassword set /vehicles/WVWXXXXXXX/controls/charging start```

works without a problem.

i dont know if this is me doing something very wrong.

**Reply by tillsteinbach** [reply]
Is that only for charging or also for climatisation?

**Reply by boredomwontgetus** [reply]
yes. its the same with climatisation for me. Works with cli and not over mqtt.

**Reply by tillsteinbach** [reply]
hmmm, I just tried to start the climatization through MQTT and it worked. Unfortunately my car is not connected to a charger right now. Thus I cannot try it out with charging.

**Reply by boredomwontgetus** [reply]
ok. ty for looking into this that quick. i'll do some more trial&error with more verbose logging on.

**Reply by boredomwontgetus** [reply]
looks like it is working now! For some reason i set ```--no-capabilities``` in the past. looks like this lead to problem. but i am not entirely sure.

**Reply by tillsteinbach** [reply]
Ok, thanks for letting me know. I will have a look if there are any conflicts when setting --no-capabilities
